File: 39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */
#include <bits/stdc++.h>

using std::vector;
using std::max;

// @lc code=start
class Solution {
public:
    // Sorting + backtracking
    // T: O(n ^ l) worst case, n := len(candidates), l := target / min(candidates)
    // S: O(l)
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());  // helps pruning
        vector<int> path;
        backtrack(path, candidates, target, res);
        return res;
    }

private:
    static void backtrack(vector<int> &path, const vector<int>& candidates, int remain, vector<vector<int>> &res) {
        if (remain == 0) {
            res.push_back(vector<int>(path));
            return;
        }
        // The same number may be chosen from candidates an unlimited number of times
        for (const int &num : candidates) {
            if (num > remain) {
                // prune
                break;
            }
            if (!path.empty() && num < path.back()) {
                // candidates sorted
                continue;
            }
            path.push_back(num);
            backtrack(path, candidates, remain - num, res);
            path.pop_back();
        }
    }
};
// @lc code=end

File: 39.combination-sum.go
/*
 * @lc app=leetcode id=39 lang=golang
 *
 * [39] Combination Sum
 */
package lc

import "sort"

// @lc code=start
func backtrack(candidates []int, path *[]int, remain int, res *[][]int) {
	if remain == 0 {
		combo := make([]int, len(*path))
		copy(combo, *path)
		*res = append(*res, combo)
	}
	// The same number may be chosen from candidates an unlimited number of times,
	// so iterate all candidates (with some pruning)
	for _, num := range candidates {
		if num > remain {
			// IMPORTANT prune: stop if exceeding target
			break
		}
		if len(*path) != 0 && num < (*path)[len(*path)-1] {
			// IMPORTANT prune: don't reuse smaller numbers
			continue
		}
		*path = append(*path, num)
		backtrack(candidates, path, remain-num, res)
		*path = (*path)[:len(*path)-1]
	}
}

// T: O(len(candidates) ^ target) worst case
// S: O(len(candidates) * target) ?
func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates) // sort numbers to allow recursion tree pruning
	var res [][]int
	var path []int
	backtrack(candidates, &path, target, &res)
	return res
}

// @lc code=end
